### PR TITLE
Add ability to load MMapper2 XML maps

### PR DESCRIPTION
**Pull request by cosmos72**
Loading back MMapper2 XML maps is **not yet finished**: exits and markers are not supported yet,
but I'd like @nschimme to have a first look at the code, which is already quite complete.

The choice to avoid comma-separated lists inside attribute values, and use instead repeated attributes,
together with some reflection-like technique, resulted in very compact and (in my opinion) elegant code.

And I'm surprised at the speed of loading big XML files:
on my system, the exported `arda.mm2xml` is ~25MB (against the 5.4MB of `arda.mm2`) yet it loads in less than two seconds :)

### CODE REVIEW

**Comment by cosmos72** [comment]
Loading `.mm2xml` maps is now complete - except for possible bugs, of course

File: src/mapstorage/XmlMapStorage.cpp [hunk 1]
```diff
@@ -27,11 +33,178 @@
...
+// ---------------------------- XmlMapStorage::Converter -----------------------
+class XmlMapStorage::Converter
+{
+public:
+    Converter();
+
+    // parse string containing an unsigned number.
+    // sets fail = true only in case of errors, otherwise fail is not modified
+    template<typename T>
+    static std::enable_if_t<std::is_integral<T>::value && std::is_unsigned<T>::value, T> //
```
Comment (nschimme) [inline src/mapstorage/XmlMapStorage.cpp:68]: Do we even need the unsigned version of toNumber()?

**Review by nschimme — CHANGES_REQUESTED** [review]

File: src/mapdata/ExitDirection.cpp [hunk 2]
```diff
@@ -120,6 +120,20 @@ const char *lowercaseDirection(const ExitDirEnum dir)
 #undef CASE
 }
 
+ExitDirEnum directionForLowercase(const QStringRef &lowcase)
```
[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapdata/ExitDirection.cpp]: Done, it now uses `std::u16string_view`.
I am not sure you will like my hand-written, non-allocating `operator==` between `std::u16string_view` and `std::string_view` in src/global/string_view_utils.cpp

File: src/mapdata/infomark.h [hunk 3]
```diff
@@ -92,6 +92,9 @@ DEFINE_ENUM_COUNT(InfoMarkClassEnum, NUM_INFOMARK_CLASSES)
     X(Coordinate, Position2, ) \
     X(int, RotationAngle, = 0)
 
+class InfoMark;
```
[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapdata/infomark.h]: I did not see the previous declaration. Removed.

File: src/mapstorage/XmlMapStorage.cpp [hunk 4]
```diff
@@ -27,11 +33,178 @@
 #include "progresscounter.h"
 #include "roomsaver.h"
 
+// ---------------------------- XmlMapStorage::Type ------------------------
+// list know enum types
+#define X_FOREACH_TYPE_ENUM(X) \
```
[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp:39]: Done.

I must admit I was surprised by the "Unity" build system #including multiple .cpp files and compiling them as a single unit, as it has some subtle side effects:
- you must write #undef of locally #defined macros at the end of .cpp files
- static data and static free functions leak into the other .cpp files

Out of curiosity, is there a document somewhere explaining the rationale behind it?

File: src/mapstorage/XmlMapStorage.cpp [hunk 5]
```diff
@@ -27,11 +33,178 @@
...
+    template<typename ENUM>
+    const QString &toString(ENUM val) const
+    {
+        static_assert(std::is_enum<ENUM>::value, "template type ENUM must be an enumeration");
+        return enumToString(enumToType(val), uint(val));
+    }
+
+private:
+    // convert an enumeration type to Type value, also used as index in enumToString[] and stringToEnum[]
+    static constexpr Type enumToType(RoomAlignEnum) { return Type::RoomAlignEnum; }
```
[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp]: Done.

File: src/mapstorage/XmlMapStorage.cpp [hunk 6]
```diff
@@ -27,11 +33,178 @@
...
+class XmlMapStorage::Converter
+{
+public:
+    Converter();
+
+    // parse string containing an unsigned number.
+    // sets fail = true only in case of errors, otherwise fail is not modified
+    template<typename T>
+    static std::enable_if_t<std::is_integral<T>::value && std::is_unsigned<T>::value, T> //
+    toNumber(const QStringRef &str, bool &fail)
```
[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp]: I have replaced `QStringRef` -> `QStringView` everywhere in XmlMapStorage.h and XmlMapStorage.cpp
Luckily there is an implicit constructor `QStringView(QStringRef)`

Where possible, I even used std::u16string_vew instead.

File: src/mapstorage/XmlMapStorage.cpp [hunk 7]
```diff
@@ -41,20 +214,396 @@ void XmlMapStorage::newData()
...
+    RoomLoadFlags loadFlags;
+    RoomMobFlags mobFlags;
+
+    while (stream.readNextStartElement() && !stream.hasError()) {
+        const QStringRef name = stream.name();
+        if (name == "align") {
+            room.setAlignType(loadEnum<RoomAlignEnum>(stream));
+        } else if (name == "contents") {
+            room.setContents(RoomContents{loadString(stream)});
+        } else if (name == "coord") {
```
[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp:344]: 1. About making room coord a required field: I have used the "be conservative in what you send, be liberal in what you accept" principle (https://en.wikipedia.org/wiki/Robustness_principle), and thus kept required fields to an absolute minimum. This is important especially when an XML map is modified manually by a user. If coordinates are missing, they will default to 0,0,0

2. My early iterations of XML map loading did not load back room coordinates, and created all 28000+ rooms at 0,0,0: exactly the case you ask about. MMapper is smart in this case, and moves the second ad further rooms at the nearest unoccupied coordinates. It becomes somewhat slow if all nearby coordinates are occupied, and it has to search a lot for a free spot.

File: src/mapstorage/XmlMapStorage.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp]: Hmm, `QStringView::toLong()` and `QStringView::toULong()` are missing in Qt < 5.15.2
I'll revert those (and only those) to QStringRef for the time being

[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp]: In the end I decided to remove QStringRef once and for all from XmlMapStorage.cpp,
and implement a conversion function from std::u16string_view to integer numbers.

File: src/mapstorage/XmlMapStorage.cpp [hunk 8]
```diff
@@ -41,20 +214,396 @@ void XmlMapStorage::newData()
...
+        if (type != "mmapper2xml") {
+            throwErrorFmt(stream,
+                          "unsupported map type=\"%1\",\nexpecting type=\"mmapper2xml\"",
+                          type);
+        }
+        const QString version = attrs.value("version").toString();
+        const CompareVersion cmp(version);
+        if (cmp.major() != 1) {
+            throwErrorFmt(stream,
+                          "unsupported map version=\"%1\",\nexpecting version=\"1.x.y\"",
```
[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp:267]: The idea was a message meaning
"unsupported map version 2.3.7: this MMapper can only load map versions 1.something"

Suggestions on how to reword it?

File: src/mapstorage/XmlMapStorage.cpp [hunk 9]
```diff
@@ -41,20 +214,396 @@ void XmlMapStorage::newData()
...
+
+// load current <exit> element
+void XmlMapStorage::loadExit(QXmlStreamReader &stream, ExitsList &exitList)
+{
+    const QXmlStreamAttributes attrs = stream.attributes();
+    const ExitDirEnum dir = directionForLowercase(attrs.value("dir"));
+    DoorFlags doorFlags;
+    ExitFlags exitFlags;
+    Exit &exit = exitList[dir];
+    exit.setDoorName(DoorName{attrs.value("doorname").toString()});
```
[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp:402]: Luckily XML as a file format is recent enough that it automatically handles encoding woes (quite) sanely.

The first line of XML files generated with "Export MMapper2 Base Map As..."  always contains the following:
```
<?xml version="1.0" encoding="UTF-8"?>
```
Users can manually change that encoding in their own saved XML files if they want,
and QXmlStreamReader should honour whatever encoding is written in the first line when loading an XML map,
automatically converting it internally to QString and/or QStringRef (which are UTF-16).

TaggedString<> and its aliases (DoorName, InfoMarkText, RoomName, RoomDesc, RoomContents, RoomNote)
automatically do the necessary (and as correct as possible) conversion from UTF-16 to Latin1 in the constructor
```
explicit TaggedString(const QString &s)
        : m_str(::toStdStringLatin1(s))
    {}
```
In summary, the code is already correct.

I tested loading an XML map containing the non-ascii characters ÀÈÌÒÙ ÁÉÍÓÚ àèìòù áéíúó
(encoded as UTF-8 in the file, since it started with encoding="UTF-8"), and they were shown correctly in the GUI.

If you write in the file some characters **not** representable in Latin1, as for example the euro sign €,
they will be converted to '?' by the `::toStdStringLatin1(s)` called by `TaggedString` constructor.

This is the best we can currently do, because MMapper2 internally uses Latin1 for most strings.
If it used UTF-8 or UTF-16 (never mind UTF-32), loading/saving any unicode character from/to XML would be perfect.

File: src/mapstorage/XmlMapStorage.cpp [hunk 10]
```diff
@@ -41,20 +214,396 @@ void XmlMapStorage::newData()
...
+    while (stream.readNextStartElement() && !stream.hasError()) {
+        const QStringRef name = stream.name();
+        if (name == "pos1") {
+            marker.setPosition1(loadCoordinate(stream));
+        } else if (name == "pos2") {
+            marker.setPosition2(loadCoordinate(stream));
+        } else if (name == "text") {
+            // load text only if type == TEXT
+            if (type == InfoMarkTypeEnum::TEXT) {
+                marker.setText(InfoMarkText{loadString(stream)});
```
[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp:505]: As above

File: src/mapstorage/XmlMapStorage.cpp (hunk 1 above)
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp]: I have refactored this code to avoid QStringView, and the conversion from string to unsigned integer is now used to load room IDs.

File: src/mapstorage/XmlMapStorage.cpp [hunk 11]
```diff
@@ -41,20 +214,396 @@ void XmlMapStorage::newData()
...
+    InfoMark &marker = deref(sharedmarker);
+
+    marker.setType(type);
+    marker.setClass(clas);
+    marker.setRotationAngle(angle);
+
+    while (stream.readNextStartElement() && !stream.hasError()) {
+        const QStringRef name = stream.name();
+        if (name == "pos1") {
+            marker.setPosition1(loadCoordinate(stream));
```
[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp:499]: What's the sanest default for mark type and mark class?
We should use that.

A mark without position currently defaults to 0,0,0 which is admittedly less useful than rooms defaulting at 0,0,0.
Still, it seems better to me than rejecting the whole map because of a missing coordinate.

File: src/mapstorage/XmlMapStorage.cpp [hunk 12]
```diff
@@ -41,20 +214,396 @@ void XmlMapStorage::newData()
...
+        m_mapData.clear();
+        return false;
+    }
+}
+
+void XmlMapStorage::loadWorld(QXmlStreamReader &stream)
+{
+    ProgressCounter &progressCounter = getProgressCounter();
+    progressCounter.reset();
+    progressCounter.increaseTotalStepsBy(32728); // just a guess
```
[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp:237]: We could add a room count to the XML file, as long as it's only used to set the progress bar range:
I don't want to require an exact count, because it would force users to count the rooms after they edited a file - a completely unnecessary chore.

But since it's only for the progress bar, maybe we could also live without the room count?

[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp:237]: [UPDATE] I found a better solution:
retrieve file size when opening the XML file, and divide `QXmlStreamReader::characterOffset()` by file size to compute the progress. I will implement it.

[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp]: Done.

File: src/mapdata/ExitDirection.h [hunk 13]
```diff
@@ -32,7 +33,12 @@ NODISCARD extern bool isNESW(ExitDirEnum dir);
 NODISCARD extern bool isUpDown(ExitDirEnum dir);
 NODISCARD extern bool isNESWUD(ExitDirEnum dir);
 NODISCARD extern ExitDirEnum opposite(ExitDirEnum in);
-NODISCARD extern const char *lowercaseDirection(ExitDirEnum dir);
+NODISCARD extern std::string_view lowercaseDirection(ExitDirEnum dir);
+NODISCARD inline const char *lowercaseDirectionC(ExitDirEnum dir)
```
Comment (nschimme) [inline src/mapdata/ExitDirection.h:37]: Why are we changing the lowest API to support a std::string_view and then wrapping it back to a const char *?

Shouldn't we just keep it as a const char * and then use a std::string_view in the one class where it was needed?

File: src/mapstorage/XmlMapStorage.cpp [hunk 14]
```diff
@@ -27,11 +35,160 @@
...
+    return empty;
+}
+
+uint XmlMapStorage::Converter::stringToEnum(Type type, const QStringView str, bool &fail) const
+{
+    const uint index = uint(type);
+    if (index < stringToEnums.size()) {
+        auto iter = stringToEnums[index].find(str);
+        if (iter != stringToEnums[index].end()) {
+            return iter.value();
```
Comment (nschimme) [inline src/mapstorage/XmlMapStorage.cpp:174]: What if we have a key collision between enums? This solution is very generic which is good from a code duplication standpoint but its brittle in its current form. I'm wondering now if a solution like what we have in `AbstractParser-Room.cpp` might be better: https://github.com/MUME/MMapper/blob/master/src/parser/AbstractParser-Room.cpp#L95-L136

File: src/mapstorage/XmlMapStorage.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (nschimme) [reply src/mapstorage/XmlMapStorage.cpp:273]: Never mind, this is fine. I misread the code.

File: src/mapstorage/XmlMapStorage.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (nschimme) [reply src/mapstorage/XmlMapStorage.cpp:328]: LGTM

File: src/mapstorage/XmlMapStorage.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (nschimme) [reply src/mapstorage/XmlMapStorage.cpp:403]: LGTM, thanks for explaining

File: src/mapstorage/XmlMapStorage.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (nschimme) [reply src/mapstorage/XmlMapStorage.cpp:500]: I think we should reject the map. The power user should edit it to fix the error rather than it silently corrupting more over time. I don't see this XML mode being something most of our normal users use so we should be more strict given the risk it brings.

**Review by nschimme — CHANGES_REQUESTED** [review]

File: src/mapdata/ExitDirection.h (hunk 13 above)
Comment (cosmos72) [reply src/mapdata/ExitDirection.h:37]: I rewrote it like that because the conversion `std::string_view` -> `const char *` is O(1): you just extracts the pointer.
Instead the conversion `const char *` -> `std::string_view` is O(n): it must call `strlen()` or equivalent.

So `std::string_view` is the better alternative to return from such a function: the alternate function returning `const char *` can be defined on top of it.

Also, I could not find a pair of names better than `lowercaseDirection()` and `lowercaseDirectionC()`:
the alternative `lowercaseDirectionStringView` or some abbreviation of it really looked ugly to me.

File: src/mapdata/ExitDirection.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (nschimme) [reply src/mapdata/ExitDirection.cpp]: Seems fine but I'm wondering if we should do this much low level stuff with UTF-16 ourselves. For instance, we convert to US-ASCII and then use something like `std::stoi()` to convert a number.

File: src/mapstorage/XmlMapStorage.cpp (hunk 14 above)
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp:174]: I don't see how you can have a key collision:

`stringToEnums` first index is the enum **type**, computed by `enumToType()` which is autogenerated using the macro `X_FOREACH_TYPE_ENUM` which assigns a different value to each type.
Attempts to call `enumToType()` with an enum not listed in `X_FOREACH_TYPE_ENUM` will fail at compile time, because the compiler will not find a match among the overloaded methods  `enumToType()`

`stringToEnums` second index is the enum UPPERCASE name, which must be unique within the enum values - otherwise the enum definition itself will not compile. And of course `QHash<QStringView, int>` is required to handle correctly **hash** collisions, just like `std::unordered_map<>` would do.

File: src/mapstorage/XmlMapStorage.cpp [hunk 15]
```diff
@@ -27,11 +35,160 @@
...
+    : enumToStrings{
+#define DECL(X) /* */ {#X},
+#define DECL_(X, ...) {#X},
+        {X_FOREACH_RoomAlignEnum(DECL)},
+        {X_FOREACH_DOOR_FLAG(DECL_)},
+        {X_FOREACH_EXIT_FLAG(DECL_)},
+        {X_FOREACH_RoomLightEnum(DECL)},
+        {X_FOREACH_ROOM_LOAD_FLAG(DECL)},
+        {X_FOREACH_INFOMARK_CLASS(DECL)},
+        {X_FOREACH_INFOMARK_CLASS(DECL)},
```
Comment (nschimme) [inline src/mapstorage/XmlMapStorage.cpp:127]: We accidentally declared this type twice

File: src/mapstorage/XmlMapStorage.cpp [hunk 16]
```diff
@@ -41,41 +198,426 @@ void XmlMapStorage::newData()
...
+
+// convert string to RoomId
+RoomId XmlMapStorage::loadRoomId(QXmlStreamReader &stream, const QStringView idstr)
+{
+    bool fail = false;
+    const RoomId id{conv.toInteger<uint32_t>(idstr, fail)};
+    // convert number back to string, and compare the two:
+    // if they differ, room ID is invalid.
+    if (fail || idstr != roomIdToString(id)) {
+        throwErrorFmt(stream, "invalid room id \"%1\"", idstr.toString());
```
[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp:372]: Ok, I'll make marker position mandatory.

About the `fail` flag:
I did not want to throw an exception from `XmlMapStorage::Converter::toInteger<T>` because it has no context about the meaning of the number it is parsing. If it threw an exception, a caller would need to catch it and throw another exception with a more descriptive message - and forgetting to do that is easy.

Passing a `bool &fail` to `XmlMapStorage::Converter::toInteger<T>` acts as a reminder for the caller "I should check what's there after `toInteger` returns.

If you really don't like as it looks now, I can change `toInteger` to throw on non-numeric input, and wrap it everywhere with `catch (/**/) { throw /**/ }`

File: src/mapstorage/XmlMapStorage.cpp (hunk 15 above)
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp:127]: Thanks for catching this. I'll fix it and add a check to ensure that `enumToStrings` initialization remains in sync with `X_FOREACH_TYPE_ENUM`

File: src/mapstorage/XmlMapStorage.cpp (hunk 14 above)
Comment (nschimme) [reply src/mapstorage/XmlMapStorage.cpp:174]: Gotcha, first level of the data structure is the type and that prevents conflicts across types where an UPPERCASE name might have a conflict (i.e. multiple UNDEFINED, etc).

File: src/mapstorage/XmlMapStorage.cpp (hunk 15 above)
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp]: Fixed. the last line above is now `{X_FOREACH_INFOMARK_TYPE(DECL)},`

File: src/mapstorage/XmlMapStorage.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp:372]: marker pos1 is now mandatory

[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp:372]: I also updated `saveMarker` to omit pos2 if it's equal to pos1,
and updated `loadMarker` to set pos2 = pos1 if the former is not present in the XML file.

File: src/mapstorage/XmlMapStorage.cpp (hunk 14 above)
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp:174]: exactly

File: src/mapdata/ExitDirection.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapdata/ExitDirection.cpp]: Yes, I think that's what most Qt code does when you compare QString and std::string, char * or QByteArray:
convert the odd-typed argument into a QString (allocating data as needed) and then compare the two QString:s.

I removed the allocation, at the cost of writing low-level code...

File: src/mapstorage/XmlMapStorage.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp:502]: Done. I wrote the comment "marker pos1 is now mandatory" in the wrong place (below).
Marker type and class are mandatory too.

Also, as it happens quite often that pos1 and pos2 are equal, I updated `saveMarker` to omit pos2 if it's equal to pos1.
And correspondigly now marker pos2 is optional in `loadMarker`, and defaults to pos1.

**Comment by cosmos72** [comment]
Let me disagree with commit https://github.com/MUME/MMapper/pull/246/commits/6d231a041be91a305457fb131d3d8105a26e4426 "pass by reference to avoid copies"

The changes pass by reference `DoorFlags`, `ExitFlags`, `RoomLoadFlags` and `RoomMobFlags` which are derived from `enums::Flags` that is a class containing a single integer, with a trivial copy constructor.
Passing them by value is expected to simply pass them in a register - I see no need to store them to the stack and pass their address, it just makes the (compiled) code more complicated.

File: src/mapstorage/XmlMapStorage.cpp [hunk 17]
```diff
@@ -155,6 +155,10 @@ XmlMapStorage::Converter::Converter()
             if (str == "UNDEFINED") {
                 str.clear(); // we never save or load the string "UNDEFINED"
             } else {
+                if (str == "EXIT") {
+                    // we save the EXIT flag inverted => invert the name too
+                    str = "NO_EXIT";
```
Comment (nschimme) [inline src/mapstorage/XmlMapStorage.cpp:160]: Is NO_EXIT even needed? What's the difference between "north" and "south" in the below example:

```
        <exit dir="north">
            <to>1918</to>
            <exitflag>NO_EXIT</exitflag>
        </exit>
        <exit dir="south"/>
```

Comment (cosmos72) [reply src/mapstorage/XmlMapStorage.cpp:160]: The `NO_EXIT` flag - which corresponds to the lack of `EXIT` flag in C++ - is used to mark exceptional exits,
which appear in red in the map.
It indicates places where you can exit in such direction due to external events, but is not an exit you can **decide** to take.
For example, the river rapids sending you to the shore just outside Lorien forest south border.

**Comment by nschimme** [comment]
LGTM
